function-service-impl/src/main/java/com/future/function/service/impl/feature/communication/chatroom/ChatroomServiceImpl.java

```diff
@@ -1,189 +1,189 @@
+import com.future.function.service.impl.helper.PageHelper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.redis.core.RedisTemplate;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Service
+public class ChatroomServiceImpl implements ChatroomService {
+
+  private final UserService userService;
+
+  private final ChatroomRepository chatroomRepository;
+
+  @Autowired
+  public ChatroomServiceImpl(
+          UserService userService,
+          ChatroomRepository chatroomRepository
+  ) {
+    this.userService = userService;
+    this.chatroomRepository = chatroomRepository;
+  }
+
+  @Override
+  public Page<Chatroom> getChatrooms(
+    String type, String userId, Pageable pageable
+  ) {
+
+    ChatroomType chatroomType = ChatroomType.fromString(type);
+    User user = userService.getUser(userId);
+    return chatroomRepository.findAllByTypeAndMembersOrderByUpdatedAtDesc(
+      chatroomType, user, pageable);
+  }
+
+  @Override
+  public Page<Chatroom> getChatroomsWithKeyword(
+    String keyword, String userId, Pageable pageable
+  ) {
+
+    return Optional.of(userId)
+      .map(userService::getUser)
+      .map(
+        user -> chatroomRepository.findAllByTitleContainingIgnoreCaseAndMembersOrderByUpdatedAtDesc(
+          keyword, user, pageable))
+      .orElseGet(() -> PageHelper.empty(pageable));
+  }
+
+  @Override
+  public Chatroom getChatroom(String chatroomId, String userId) {
+
+    return Optional.ofNullable(chatroomId)
+      .map(chatroomRepository::findOne)
+      .map(chatroom -> this.validateAuthorization(chatroom, userId))
+      .orElseThrow(() -> new NotFoundException("Get Chatroom Not Found"));
+  }
+
+  @Override
+  public Chatroom createChatroom(Chatroom chatroom) {
+
+    if (chatroom.getType()
+      .equals(ChatroomType.PRIVATE)) {
+      List<String> memberIds = chatroom.getMembers()
+        .stream()
+        .map(User::getId)
+        .collect(Collectors.toList());
+
+      List<User> members = chatroom.getMembers()
+        .stream()
+        .map(member -> userService.getUser(member.getId()))
+        .collect(Collectors.toList());
+
+      List<Chatroom> filteredChatrooms =
+        chatroomRepository.findAllByMembersContaining(members)
+          .stream()
+          .filter(room -> room.getType()
+            .equals(ChatroomType.PRIVATE))
+          .filter(room -> isMemberIdsInChatroom(memberIds, room))
+          .collect(Collectors.toList());
+
+      if (filteredChatrooms.size() > 0) {
+        return filteredChatrooms.get(0);
+      }
+    }
+
+    return Optional.of(chatroom)
+      .map(this::setMembers)
+      .map(this::setChatroomName)
+      .map(chatroomRepository::save)
+      .orElseThrow(UnsupportedOperationException::new);
```

Comment:
consider to giving a message to exception to easier debugging

Reply:
it is not a new code anyway, but i don't know somehow it showed as addition in github